File: Shared/Model/Outline.swift
import Foundation

/// A heading in a page. The outline lists these so a long note can be jumped through.
nonisolated struct Heading: Identifiable, Equatable, Sendable {
    /// Its place among the headings, which is also its place in reading mode.
    let id: Int
    let level: Int
    let title: String
    /// Where the heading's line starts in the page's text, counted the way text views count.
    let offset: Int
}

/// Where the outline wants the page to go. The id makes a second jump to the same heading count.
nonisolated struct Jump: Equatable, Sendable {
    let offset: Int
    let heading: Int
    let id = UUID()
}

enum Outline {
    private static let heading = /^(#{1,6})[ \t]+(.+?)[ \t]*#*[ \t]*$/

    /// The headings of a page, in the order they appear. Headings inside code blocks don't count.
    nonisolated static func headings(in text: String) -> [Heading] {
        var headings: [Heading] = []
        var offset = 0
        var insideCode = false

        for line in text.split(separator: "\n", omittingEmptySubsequences: false) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.hasPrefix("```") || trimmed.hasPrefix("~~~") {
                insideCode.toggle()
            } else if !insideCode, let match = trimmed.firstMatch(of: heading) {
                headings.append(Heading(
                    id: headings.count,
                    level: match.output.1.count,
                    title: String(match.output.2),
                    offset: offset
                ))
            }
            offset += line.utf16.count + 1
        }
        return headings
    }
}

File: windows/Blad.Core/Markdown/Outline.cs
using System.Text.RegularExpressions;

namespace Blad.Core.Markdown;

/// <summary>A heading in a page. <see cref="Index"/> is its place among the headings, which is also
/// its place in reading mode; <see cref="Offset"/> is where its line starts in the text.</summary>
public sealed record Heading(int Index, int Level, string Title, int Offset);

/// <summary>The headings of a page, for the outline that jumps through a long note.</summary>
public static partial class Outline
{
    [GeneratedRegex(@"^(#{1,6})[ \t]+(.+?)[ \t]*#*[ \t]*$")]
    private static partial Regex HeadingLine();

    /// <summary>The headings in the order they appear. Headings inside code blocks don't count.
    /// Offsets count one character per line break, the way the editor's text box does.</summary>
    public static IReadOnlyList<Heading> Headings(string text)
    {
        var headings = new List<Heading>();
        var insideCode = false;
        var offset = 0;

        foreach (var line in MarkdownParser.SplitLines(text))
        {
            var trimmed = line.Trim();
            if (trimmed.StartsWith("```") || trimmed.StartsWith("~~~"))
            {
                insideCode = !insideCode;
            }
            else if (!insideCode && HeadingLine().Match(trimmed) is { Success: true } match)
            {
                headings.Add(new Heading(headings.Count, match.Groups[1].Length, match.Groups[2].Value, offset));
            }
            offset += line.Length + 1;
        }
        return headings;
    }
}
